File: swift/performance_problems/time_nanos.swift
struct Nano : CustomStringConvertible {
	var value: Int64 = 0
	init(_ x: Int64) { self.value = x }
	var description: String { return "\(self.value)" }
}

let u_nano = Nano(1)
let u_second = Nano(1_000_000_00)
let u_minute = u_second * 60

func *(lhs: Nano, rhs: Int) -> Nano { return Nano(lhs.value * Int64(rhs)); }
func *(lhs: Int, rhs: Nano) -> Nano { return Nano(Int64(lhs) * rhs.value); }
func +(lhs: Nano, rhs: String) -> String { return lhs.description + rhs; }
func +(lhs: String, rhs: Nano) -> String { return lhs + rhs.description; }
func +(lhs: Nano, rhs: Nano) -> Nano { return Nano(lhs.value + rhs.value); }

extension Int {
	var ns: Nano { return Nano(Int64(self)) }
	var s: Nano { return self * u_second }
	var i: Nano { return self * u_minute }
}

func test_seconds() {
	print("Testing second operations:\n")
	print(u_minute + u_second + Nano(500) + " = " + 1.i + 1.s + 500.ns)
	print((u_minute + u_second + Nano(500)) + " = " + (1.i + 1.s + 500.ns))
}

File: swift/performance_problems/time_nanos.nim
type Nano* = distinct int64

proc `+`*(x, y: Nano): Nano {.borrow.}
proc `*`*(x: Nano, y: int64): Nano = Nano(int64(x) * y)
proc `*`*(x: int64, y: Nano): Nano = Nano(x * int64(y))

const
  u_nano* = Nano(1)
  u_second* = Nano(1_000_000_000)
  u_minute* = u_second * 60

proc ns*(x: int64): Nano {.inline.} = Nano(x)
proc s*(x: int64): Nano {.inline.} = x * u_second
proc i*(x: int64): Nano {.inline.} = x * u_minute

proc `$`*(x: Nano): string = $(int64(x))
proc `&`*(x: Nano, y: string): string = $x & y
proc `&`*(x: string, y: Nano): string = x & $y


proc test_seconds*() =
  echo "Testing second operations:\n"
  echo u_minute + u_second + Nano(500), " = ", 1.i + 1.s + 500.ns
  echo u_minute + u_second + Nano(500) & " = " & 1.i + 1.s + 500.ns

test_seconds()
